src/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.ts

```diff
@@ -184,15 +374,80 @@ g.test('source_imageBitmap,state')
     if (closed) imageBitmap.close();
 
     t.runTest(
-      { imageBitmap },
+      { source: imageBitmap },
       { texture: dstTexture },
       copySize,
       true, // No validation errors.
       closed ? 'InvalidStateError' : ''
     );
   });
 
+g.test('source_offscreenCanvas,state')
+  .desc(
+    `
+  Test OffscreenCanvas as source image in state [valid, detached].
+
+  Transfer OffsreenCanvas to worker will detach the OffscreenCanvas
+  in main thread.
+  
+  Check whether 'InvalidStateError' is generated when OffscreenCanvas is
+  detached.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('detached', [false, true])
+      .beginSubcases()
+      .combine('copySize', [
+        { width: 0, height: 0, depthOrArrayLayers: 0 },
+        { width: 1, height: 1, depthOrArrayLayers: 1 },
+      ])
+  )
+  .fn(async t => {
+    const { detached, copySize } = t.params;
+    const offscreenCanvas = t.getOffscreenCanvas(1, 1);
+    const dstTexture = t.device.createTexture({
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      format: 'bgra8unorm',
+      usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    if (detached) {
+      const workerCode =
+        "self.onmessage = function(e) {e.data.canvas.getContext('2d'); poseMessage('get 2d context');};";
+      const blob = new Blob([workerCode], { type: 'text/javascript' });
+      const url = URL.createObjectURL(blob);
+      const worker = new Worker(url);
+      worker.postMessage({ canvas: offscreenCanvas }, [offscreenCanvas]);
+      worker.onmessage = function (e) {
```

Comment:
You don't have to wait for the message before testing InvalidStateError. The OffscreenCanvas becomes detached instantly inside postMessage.

Reply:
If I don't wait this, I'll get another error about the source image doesn't have correct context. Maybe in impl we should change the validation orders?

Reply:
Ohh, I see that you are creating a context on the offscreen canvas on the receiving thread.

Yes, that shouldn't be necessary. In the [html spec](https://html.spec.whatwg.org/multipage/canvas.html#:~:text=OffscreenCanvas%20objects%20are%20transferable.%20Their%20transfer%20steps%2C%20given%20value%20and%20dataHolder%2C%20are%20as%20follows%3A), the "context mode" of the source OffscreenCanvas object is set to "detached" inside postMessage, so, regardless of whether the receiver created a context, it should fail in the same way. It sounds like our implementation doesn't mirror the spec exactly, so I guess it's best to test both cases: one where the receiver creates a context prior to runTest, and one where it doesn't.

Reply:
However you can still do that without launching a worker: just grab the message from port2 (on the same thread instead of a worker) and getContext on that.

Reply:
< Yes, that shouldn't be necessary. In the html spec, the "context mode" of the source OffscreenCanvas object is set to 
<" detached" inside postMessage, so, regardless of whether the receiver created a context, it should fail in the same way.  
< sounds like our implementation doesn't mirror the spec exactly.
 Yes, I think our implementation checks the context first and then check the detached state, which I think it's not correct.

Reply:
< one where the receiver creates a context prior to runTest, and one where it doesn't.
Implement the second one. And blocks by  case return webgpu device to pool before callback(onmessage) has been hit.